include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-    
-    return retval;
-  }
-  // overloads for surface_interactions
-  template<class ObserverRayDirInfo>
-  static LightSample<RayDirInfo> create(const float3 L, const surface_interactions::Isotropic<ObserverRayDirInfo> interaction)
-  {
-    const float3 V = interaction.V.getDirection();
-    const float VdotL = dot(V,L);
-    return create(L,VdotL,interaction.N);
-  }
-  template<class ObserverRayDirInfo>
-  static LightSample<RayDirInfo> create(const float3 L, const surface_interactions::Anisotropic<ObserverRayDirInfo> interaction)
-  {
-    const float3 V = interaction.V.getDirection();
-    const float VdotL = dot(V,L);
-    return create(L,VdotL,interaction.T,interaction.B,interaction.N);
-  }
-  //
-  float3 getTangentSpaceL()
-  {
-    return float3(TdotL,BdotL,NdotL);
-  }
-
-  RayDirInfo L;
-  float VdotL;
-
-  float TdotL; 
-  float BdotL;
-  float NdotL;
-  float NdotL2;
+    NBL_CONSTEXPR_STATIC_INLINE bool is_aniso = true;
 };
 
 
-//
-struct IsotropicMicrofacetCache
+#define NBL_CONCEPT_NAME Sample
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (_sample, T)
+#define NBL_CONCEPT_PARAM_1 (iso, typename T::isotropic_type)
+#define NBL_CONCEPT_PARAM_2 (aniso, typename T::anisotropic_type)
+#define NBL_CONCEPT_PARAM_3 (rdirinfo, typename T::ray_dir_info_type)
+#define NBL_CONCEPT_PARAM_4 (pV, typename T::vector3_type)
+#define NBL_CONCEPT_PARAM_5 (frame, typename T::matrix3x3_type)
+#define NBL_CONCEPT_PARAM_6 (pVdotL, typename T::scalar_type)
+NBL_CONCEPT_BEGIN(7)
+#define _sample NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define iso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define aniso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+#define rdirinfo NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_3
+#define pV NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_4
+#define frame NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_5
+#define pVdotL NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_6
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::ray_dir_info_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::isotropic_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::vector3_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::matrix3x3_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((_sample.getL()), ::nbl::hlsl::is_same_v, typename T::ray_dir_info_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((_sample.getVdotL()), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((_sample.getTdotL()), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((_sample.getBdotL()), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((_sample.getNdotL()), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((_sample.getNdotL2()), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::createFromTangentSpace(pV,rdirinfo,frame)), ::nbl::hlsl::is_same_v, T))
```

Comment:
for the same reason we backed out of `frisvad` returning a mat3x3, we should take the frame as T,B,N vectors instead of a matrix

which I guess is your `T::create(rdirinfo,pVdotL,pV,pV,pV)` constructor ?

Reply:
unless you ofc call the argument `fromTangent` then its clear that its a matrix for transfroming out of tangent space

Reply:
Yeah `createFromTangentSpace` directly takes the frame produced by `interaction.getFromTangentSpace`